src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +72,215 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
 {
-  // Mean and variance over the entire training set will be used to compute
-  // the forward pass when deterministic is set to true.
-  if (deterministic)
-  {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
-  }
-  else
+  const size_t batchSize = input.n_cols;
+
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
+
+  // We will calculate minibatch norm on each channel / feature map.
+  if (!deterministic)
   {
-    mean = arma::mean(input, 1);
-    variance = arma::var(input, 1, 1);
+    Log::Assert(input.n_cols > 0, "Empty input. Batch size must be \
+        greater than zero.");
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+    Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+        by feature maps.");
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; ++i)
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
```

Comment:
Do you think we could remove this special case from here and just use the convolutional layer part for computing the batchnorm for the linear layer as well. This would be the case where there is only one 1 pixel in the image. Let me know your thoughts.

Reply:
I think that should work. I'll make the changes. Thanks.

Reply:
Or maybe we should keep this as optimization to the linear layer.

Reply:
I think we should leave this as if for now since it would work faster for linear layer. Let me know what you think.

Reply:
I think so, let's leave it as it is for now, if needed we can open an issue for discussion.